sylvia-derive/src/utils.rs

```diff
@@ -1,14 +1,43 @@
+use proc_macro2::Span;
 use proc_macro_error::emit_error;
 use syn::spanned::Spanned;
 use syn::visit::Visit;
 use syn::{
-    FnArg, GenericArgument, GenericParam, PathArguments, PathSegment, ReturnType, Signature, Type,
-    WhereClause, WherePredicate,
+    Attribute, FnArg, GenericArgument, GenericParam, ImplItem, ItemImpl, ItemTrait, PathArguments,
+    PathSegment, ReturnType, Signature, TraitItem, Type, WhereClause, WherePredicate,
 };
 
 use crate::check_generics::CheckGenerics;
 use crate::message::MsgField;
 
+pub type ItemsIterator<'a> =
+    Box<dyn Iterator<Item = (&'a Vec<Attribute>, &'a Signature, Span)> + 'a>;
+
+/// Trait for extracting attributes and signature of the methods from `ItemImpl` and `ItemTrait`
+/// In most cases these two parameters are being used for preprocessing
+/// so to unify the logic we can use this trait
+pub trait Items<'a> {
+    fn items(&'a self) -> ItemsIterator;
+}
```

Comment:
I really like the implementation, I don't like naming. `Items` is a very generic name; it is not very clear what we are talking about. I know that here it is either `TraitItem` or `ImplItem`, but it is only because I know syn pretty well, but it doesn't really spell out what is its role in the codebase. Especially since the iterator inner type is also called... `Item`, which makes the name very confusing.

The `Item` in our case is something like `VariantDesc`, or maybe `MessageDesc` - something along those lines (you can come up with something with more sense, but something describing the role of it in the code).

Therefore the first thing I would do is to create the proper type for an "item". like `struct VariantDesc { ... }` containing fields for your tuple. This way, you never need to remember the field's order, and it is a generally better idea.

Additionally, as you keep `Span` in the type, it is probably a good idea to implement [syn::spanned::Spanned](https://docs.rs/syn/latest/syn/spanned/trait.Spanned.html) on the type - it looks natural.

The question here is if the items on the struct should be public. The `Span` for sure not - you should access it via the syn trait you implement on this.  However, you can either consider the `VariantDesc` type as just a data passing type and keep fields public, or you can figure out what kind of context it is used for and maybe provide proper functions on it. I would always prefer the second unless it ends up just implementing getters for its fields - but even in such a case keeping proper harmonization ends up being beneficial (as you almost always don't want to expose easy ways to modify fields after they are generated). Maybe I will give you a hint of how to work on it later and comment again.

Then I would rename the `Items` type to something like `IntoVariantDescs`, but implement it slightly differently. Here you used just `Box<dyn Iterator<...>>`, which is ok, but it is not very flexible - sometimes you want to provide the exact iterator type in implementation to make it better runtime efficient.

Therefore it should have an associated type named `Iter: Iterator<Item=VariantDesc>`, which is then used as the return value. You will still make it `Box<dyn Iterator<...>>` (or even use your alias, which is fine) in this context, but it gives you more flexibility and also prepares you for the upcoming [#63063](https://github.com/rust-lang/rust/issues/63063), when you will just change `Box<dyn Iterator<...>>` into `impl Iterator<...>` and you will get monopolization for free. In general you are creating the alternative [IntoIterator](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html).

The tempting idea would be to not even create this trait, but instead, just implement the `IntoIterator` trait directly on `ItemImpl` and `ItemTrait`, but you cannot do it because of the orphan rule (and it would be also less verbose which is not ideal).